### PR TITLE
[hail] Rewrite PType.subsetTo to avoid using `canonical`, which is wrong

### PR DESCRIPTION
**Pull request by tpoterba**
Stacked on #7869

### CODE REVIEW

**Comment by tpoterba** [comment]
parent PR is in, ready for review.

File: hail/src/main/scala/is/hail/expr/types/physical/PType.scala [hunk 1]
```diff
@@ -275,8 +275,29 @@ abstract class PType extends Serializable with Requiredness {
...
+        PCanonicalSet(e.subsetTo(ts.elementType), r)
+      case PCanonicalDict(k, v, r) =>
+        val td = t.asInstanceOf[TDict]
+        PCanonicalDict(k.subsetTo(td.keyType), v.subsetTo(td.valueType), r)
+      case PCanonicalInterval(p, r) =>
+        val ti = t.asInstanceOf[TInterval]
+        PCanonicalInterval(p.subsetTo(ti.pointType), r)
+      case _ =>
+        assert(virtualType == t)
+        this
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: Should this be t.cop, since copies are returned for other branches?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: no, there's no reason to copy unnecessarily. PTypes are functional objects (no methods mutate them), so we should always avoid reconstructing if it's an option!

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: Sure, as long as we're consistent in that this won't be a problem

**Pull request merged**